docs/routing.md

```diff
@@ -0,0 +1,266 @@
+
+#### Diagram
+
+```mermaid
+flowchart TD
+    checkRoute --> CheckSourceMatch{check source match}
+    CheckSourceMatch -->|true| ApplyOverrides[apply overrides]
+    CheckSourceMatch -->|false| EndRoutingSkip[`skip`]
+    ApplyOverrides --> ApplyInternationalization[apply internationalization\nredirects]
+    ApplyInternationalization --> HasMiddleware{has middleware}
+    HasMiddleware --> |yes| RunMiddlewareResult{run middleware}
+    HasMiddleware --> |no| ApplyHeadersStatusAndDestination[apply route headers\napply status code\napply destination]
+    RunMiddlewareResult --> |success| ApplyMiddlewareHeaders[apply middleware\nheaders]
+    RunMiddlewareResult --> |error| EndRoutingError[`error`]
+    ApplyMiddlewareHeaders --> ApplyHeadersStatusAndDestination
+    ApplyHeadersStatusAndDestination --> ShouldCheck{should check filesystem}
+    ShouldCheck --> |yes| PathsEqual{new path is equal}
+    PathsEqual --> |yes\n+\nphase = miss| Return404[set status to 404] --> ShouldContinue
+    PathsEqual --> |yes\n+\nphase != miss| checkNextPhase[check next phase]
+    PathsEqual --> |false| checkNonePhase[check `none` phase]
+    ShouldCheck --> |false| ShouldContinue{should continue\n and has no\nredirect}
+    ShouldContinue --> |true| ReturnNext[`next`]
+    ShouldContinue --> |false| ReturnDone[`done`]
+```
+
+### Serving Responses
+
+After the routing process has reached the end and found a match, the final thing to do is serve the response. First, it is necessary to determine whether a redirect should occur, based on the location header, so that it can be returned before running a function or fetching a static asset.
+
+If a redirect should occur, we simply return the redirect response.
+
+Otherwise, we fetch the record from the build output for the matched path. If we failed to find a record, it results in a 404 response. With the record, we can then run any functions, or fetch any static assets, depending on the type. Before running the function, the final search parameters are applied to the request object.
+
+After the response object has been created, we apply any headers and status codes that were specified by the source routes.
+
+Finally, we return the response.
+
+#### Diagram
+
+```mermaid
+flowchart TD
+    serve --> CheckRedirect{check for location header}
+    CheckRedirect --> |true| ReturnRedirect[return redirect response]
+    CheckRedirect --> |false| FetchItemFromOutput[fetch item from build output]
+    FetchItemFromOutput --> ApplySearchParams[apply search params to request]
+    ApplySearchParams --> RetrieveResponseForItem{fetch or run\nitem response}
+    RetrieveResponseForItem --> |success| ApplyHeadersAndStatus[apply headers + status]
+    RetrieveResponseForItem --> |error| ReturnError[return error response]
+    ApplyHeadersAndStatus --> ReturnResponse[return final response object]
+```
```

Comment:
```suggestion
```mermaid
flowchart TD
    serve --> checkRedirect{check for\n`location`\nheader}
    checkRedirect --> |found| returnRedirect[return redirect response]
    checkRedirect --> |not found| fetchItemFromOutput[fetch item from build output]
    fetchItemFromOutput --> applySearchParams[apply search params to request]
    applySearchParams --> retrieveResponseForItem{fetch item\nresponse}
    retrieveResponseForItem --> |success| applyHeadersAndStatus[apply headers + status]
    retrieveResponseForItem --> |error| returnError[return error response]
    applyHeadersAndStatus --> returnResponse[return final response object]
```
```